File: python3/minimum_operations_to_reduce_x_to_zero.py
"""
Space   : O(n)
Time    : O(n)
"""


class Solution:
    def minOperations(self, nums: List[int], x: int) -> int:
        longest = 0
        n = len(nums)
        k = sum(nums) - x
        d = {}
        d[0] = 0

        if k == 0:
            return len(nums)

        i, count = 0, 0
        while i < n:
            count += nums[i]
            d[count] = i+1

            if count - k in d:
                longest = max(longest, i - d[count-k]+1)

            i += 1

        if longest > 0:
            return n - longest

        return -1

File: golang/minimum_operations_to_reduce_x_to_zero.go
package main

// Space	: O(n)
// Time		: O(n)

func minOperations(nums []int, x int) int {
	longest := 0
	allSummInArray := 0
	for _, num := range nums {
		allSummInArray += num
	}

	if allSummInArray < x {
		return -1
	} else if allSummInArray == x {
		return len(nums)
	}

	sumTotal := 0
	left, right := 0, 0

	for left < len(nums) {
		sumTotal += nums[left]
		left++
		for sumTotal > allSummInArray-x {
			sumTotal -= nums[right]
			right++
		}

		if sumTotal == allSummInArray-x {
			longest = Max(longest, left-right+1)
		}
	}

	if longest == 0 {
		return -1
	}

	return len(nums) - longest + 1
}
